### PR TITLE
[IMP] website_forum: search and filter tags

### PR DESCRIPTION
**Pull request by flch-odoo**
The first commit is a bugfix that will come from earlier versions.

The second commit enables user to

* Track, show, and use the forum's
  * Most commonly used tags
  * Unused tags (for the moderators)
* Search tags
* Get an improved "no results" experience with tags

Task-3349375

### CODE REVIEW

File: addons/website_forum/controllers/website_forum.py [hunk 1]
```diff
@@ -179,31 +180,53 @@ def tag_read(self, forum_id, query='', limit=25, **post):
...
+                'forum_tags_only', search, limit=None, order=order, options={'forum': forum},
+            )
+            tags = details[0].get('results', tags)
+
+        if filters in ('unused', 'most_used'):
+            filter_tags = forum.tag_most_used_ids if filters == 'most_used' else forum.tag_unused_ids
+            tags = tags & filter_tags if tags else filter_tags
+
+        elif filters in ('all', 'followed'):
+            domain = ['&', ('forum_id', '=', forum.id), ('posts_count', '>', 0)]
```
Comment (std-odoo) [inline addons/website_forum/controllers/website_forum.py]:
```suggestion
            domain = [('forum_id', '=', forum.id), ('posts_count', '>', 0)]
```

File: addons/website_forum/models/forum_forum.py [hunk 2]
```diff
@@ -131,11 +134,50 @@ def _get_default_welcome_message(self):
...
+            return
+
+        query = """
+            SELECT id, forum_id, posts_count
+              FROM forum_tag
+             WHERE id IN %(tag_ids)s
+          ORDER BY forum_id, posts_count DESC, name
+       """
+        self.env.cr.execute(query, {'tag_ids': tuple(self.tag_ids.ids)})
+        tags_data = self.env.cr.fetchall()
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: why not doing a `search_read` ?

File: addons/website_forum/models/forum_forum.py [hunk 3]
```diff
@@ -256,10 +298,13 @@ def _tag_to_write_vals(self, tags=''):
...
-        """ get set of first letter of forum tags """
-        tags = self.env['forum.tag'].search([('forum_id', '=', self.id), ('posts_count', '>', 0)])
-        return sorted(set([tag.name[0].upper() for tag in tags if len(tag.name)]))
+    def get_tags_first_char(self, tags=None):
+        """Get set of first letter of forum tags.
+
+        :param tags: tags recordset to further filter forum's tags that are also in these tags.
+        """
+        tag_ids = self.tag_ids if not tags else (self.tag_ids & tags)
+        return sorted(set([tag.name[0].upper() for tag in tag_ids if len(tag.name)]))
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]:
```suggestion
        return sorted({tag.name[0].upper() for tag in tag_ids if len(tag.name)})
```

File: addons/website_forum/models/forum_forum.py [hunk 4]
```diff
@@ -131,11 +134,50 @@ def _get_default_welcome_message(self):
...
+    def _compute_tag_ids_usage(self):
+        if not self.tag_ids:
+            self.tag_unused_ids = self.tag_most_used_ids = False
+            return
+
+        query = """
+            SELECT id, forum_id, posts_count
+              FROM forum_tag
+             WHERE id IN %(tag_ids)s
+          ORDER BY forum_id, posts_count DESC, name
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: just to make it deterministic

```suggestion
          ORDER BY forum_id, posts_count DESC, name, id
```

File: addons/website_forum/models/forum_forum.py [hunk 5]
```diff
@@ -131,11 +134,50 @@ def _get_default_welcome_message(self):
...
+        current_forum_id = tags_data[0][1]
+        current_most_used_count = 0
+        forum_tags = defaultdict(lambda: {'most_used_ids': [], 'unused_ids': []})
+
+        for tag_id, tag_forum_id, posts_count in tags_data:
+            if tag_forum_id != current_forum_id:
+                current_most_used_count = 0
+                current_forum_id = tag_forum_id
+            else:
+                if not posts_count:  # Could be 0 or None
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: you can do an `elif` here and save one indentation

File: addons/website_forum/models/forum_forum.py [hunk 6]
```diff
@@ -256,10 +298,13 @@ def _tag_to_write_vals(self, tags=''):
         post_tags.insert(0, [6, 0, existing_keep])
         return post_tags
 
-    def get_tags_first_char(self):
-        """ get set of first letter of forum tags """
-        tags = self.env['forum.tag'].search([('forum_id', '=', self.id), ('posts_count', '>', 0)])
-        return sorted(set([tag.name[0].upper() for tag in tags if len(tag.name)]))
+    def get_tags_first_char(self, tags=None):
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: note that the method can be private (maybe the opportunity to do it)

File: addons/website_forum/views/forum_forum_templates.xml [hunk 7]
```diff
@@ -225,14 +235,16 @@
         </div>
         <t t-else="">
             <t t-set="tag_filter" t-value="keep_query('filters') and keep_query('filters').split('=')[1] or ''"/>
+            <t t-set="go_back_url" t-valuef="/forum/#{ slug(forum) }/tag"/>
+            <t t-set="record_name_plural" t-value="'tags'"/>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates.xml]: it won't be translated no ? (same for posts, etc)

File: addons/website_forum/tests/test_forum_tags.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                'tag_ids': [tag_id.id]
+            }
+                for _ in range(post_count)
+            ]
+        self.env['forum.post'].create(vals_list)
+        self.env['forum.tag'].flush_model()
+
+        self.assertEqual(
+            set(self.base_forum.tag_most_used_ids.ids),
+            set([tag_id.id for tag_id, _ in posts_per_tag[-MOST_USED_TAGS_COUNT:]])
```
Comment (std-odoo) [inline addons/website_forum/tests/test_forum_tags.py]: same here about set comprehension `set([...]) -> {...}`

File: addons/website_forum/tests/test_forum_tags.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        vals_list = []
+        for tag_id, post_count in posts_per_tag:
+            vals_list += [{
+                'forum_id': self.base_forum.id,
+                'name': 'A post',
+                'content': 'A content',
+                'tag_ids': [tag_id.id]
+            }
+                for _ in range(post_count)
+            ]
```
Comment (std-odoo) [inline addons/website_forum/tests/test_forum_tags.py]:
```suggestion
        vals_list = [
            {
                'forum_id': self.base_forum.id,
                'name': 'A post',
                'content': 'A content',
                'tag_ids': [tag_id.id]
            }
            for tag_id, post_count in posts_per_tag:
            for __ in range(post_count)
        ]
```

File: addons/website_forum/tests/test_forum_tags.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        self.assertEqual(self.forum.tag_unused_ids, self.forum.tag_ids - used_tag)
+
+        self.assertEqual(self.base_forum.tag_most_used_ids, self.base_forum.tag_ids[:MOST_USED_TAGS_COUNT])
+        self.assertEqual(self.base_forum.tag_unused_ids, self.env['forum.tag'])
+
+    def test_tags_counts_most_used(self):
+        posts_per_tag = [
+            (tag_id, post_count)
+            for tag_id, post_count
+            in zip(self.base_forum.tag_ids, list(range(2, 9)))
```
Comment (std-odoo) [inline addons/website_forum/tests/test_forum_tags.py]:
```suggestion
            in zip(self.base_forum.tag_ids, range(2, 9))
```

File: addons/website_forum/tests/test_forum_tags.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        })
+        new_posts |= self.env['forum.post'].create([
+            {
+                'forum_id': self.base_forum.id,
+                'name': 'A post',
+                'content': 'A content',
+                'tag_ids': [tag_id.id]
+            }
+            for tag_id in self.base_forum.tag_ids
+        ])
```
Comment (std-odoo) [inline addons/website_forum/tests/test_forum_tags.py]: you can give a list of dict to the create method instead

**Review by std-odoo — COMMENTED** [review]
@flch-odoo Hi :) 

Some comments while passing :) 

Cheers

File: addons/website_forum/models/forum_forum.py (hunk 3 above)
Comment (flch-odoo) [reply addons/website_forum/models/forum_forum.py]: Yep, let's clean as well :)

File: addons/website_forum/tests/test_forum_tags.py (hunk 11 above)
Comment (flch-odoo) [reply addons/website_forum/tests/test_forum_tags.py]: As I did but with all records? (I did that change already)

File: addons/website_forum/models/forum_forum.py (hunk 2 above)
Comment (flch-odoo) [reply addons/website_forum/models/forum_forum.py]: Turns out, no valid reason.

**Review by flch-odoo — COMMENTED** [review]
Thank you, @std-odoo !

File: addons/website_forum/models/forum_tag.py [hunk 12]
```diff
@@ -26,10 +32,41 @@ def _compute_posts_count(self):
...
+        search_fields = ['name']
+        fetch_fields = ['id', 'name', 'website_url']
+        mapping = {
+            'name': {'name': 'name', 'type': 'text', 'match': True},
+            'website_url': {'name': 'website_url', 'type': 'text', 'truncate': False},
+        }
+        base_domain = []
+        if forum := options.get("forum"):
+            forum_ids = (unslug(forum)[1], ) if isinstance(forum, str) else forum.ids
+            base_domain = [[('forum_id', 'in', forum_ids)]]
```
Comment (std-odoo) [inline addons/website_forum/models/forum_tag.py:62]: why not that ? (doesn't look like a valid domain)
```suggestion
            base_domain = [('forum_id', 'in', forum_ids)]
```

File: addons/website_forum/tests/test_forum_tag.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+                'forum_id': forum_id.id,
+                'name': f'Test Tag {tag_idx}'
+            }
+            for forum_id in forums
+            for tag_idx in range(1, 8)
+        ])
+
+    def test_tags_counts_unused(self):
+        used_tag = self.forum.tag_ids[5]
+        self.env['forum.post'].create(
```
Comment (std-odoo) [inline addons/website_forum/tests/test_forum_tag.py:35]: if you want to optimize it even further, you can write
```py
from itertools import zip_longest

self.env['forum.post'].create([
    {'forum_id': forum.id, 'name': 'A post', 'content': 'A content', 'tag_ids': tag}
    for tag, forum in zip_longest(used_tag | self.base_forum.tag_ids, self.forum, fillvalue=self.base_forum)
])
```

(so the name, content, etc are not repeated, current version is good enough if you prefer it, it's chipotage at this point)

File: addons/website_forum/models/forum_forum.py [hunk 14]
```diff
@@ -131,11 +135,46 @@ def _get_default_welcome_message(self):
...
+        tags_data = self.env['forum.tag'].search_read(
+            [('id', 'in', self.tag_ids.ids)],
+            fields=['id', 'forum_id', 'posts_count'],
+            order='forum_id, posts_count DESC, name, id',
+        )
+        current_forum_id = tags_data[0]['forum_id'][0]
+        current_most_used_count = 0
+        forum_tags = defaultdict(lambda: {'most_used_ids': [], 'unused_ids': []})
+
+        for tag_data in tags_data:
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: In the tests, this method is never called in batch, but the `order=forum_id` is important for the logic you implemented

screenshot_spect


Would it be possible to call something like this in your tests
```py
(self.forum | self.base_forum).tag_most_used_ids
```

with a comment saying that we do that to call the compute in batch

btw the test `test_tags_counts_unused` crash when I call in batch (before calling on single record to get the value), so I think there's a bug in the compute method

File: addons/website_forum/models/forum_forum.py [hunk 15]
```diff
@@ -131,11 +135,46 @@ def _get_default_welcome_message(self):
...
+            tag_id, tag_forum_id, posts_count = itemgetter('id', 'forum_id', 'posts_count')(tag_data)
+            if tag_forum_id[0] != current_forum_id:
+                current_most_used_count = 0
+                current_forum_id = tag_forum_id[0]
+            elif not posts_count:  # Could be 0 or None
+                forum_tags[current_forum_id]['unused_ids'].append(tag_id)
+            elif current_most_used_count < MOST_USED_TAGS_COUNT:
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: you don't need the variable `current_most_used_count`, you can just use `len(forum_tags[current_forum_id]['most_used_ids'])`

File: addons/website_forum/models/forum_forum.py [hunk 16]
```diff
@@ -131,11 +135,46 @@ def _get_default_welcome_message(self):
...
             forum.can_moderate = self.env.user.karma > forum.karma_moderate
 
+    @api.depends('post_ids', 'post_ids.tag_ids', 'post_ids.tag_ids.posts_count', 'tag_ids')
+    def _compute_tag_ids_usage(self):
+        if not self.tag_ids:
+            self.tag_unused_ids = self.tag_most_used_ids = False
+            return
+
+        tags_data = self.env['forum.tag'].search_read(
+            [('id', 'in', self.tag_ids.ids)],
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: can we do that ? (so we have a smaller domain)
```suggestion
            [('forum_id', 'in', self.ids)],
```

File: addons/website_forum/tests/test_forum_tag.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+            {'forum_id': self.base_forum.id, 'name': 'A post', 'content': 'A content', 'tag_ids': [tag_id.id]}
+            for tag_id, post_count in posts_per_tag
+            for __ in range(post_count)
+        ]
+        self.env['forum.post'].create(vals_list)
+        self.env['forum.tag'].flush_model()
+
+        self.assertEqual(
+            set(self.base_forum.tag_most_used_ids.ids),
+            set(tag_id.id for tag_id, _ in posts_per_tag[-MOST_USED_TAGS_COUNT:])
```
Comment (std-odoo) [inline addons/website_forum/tests/test_forum_tag.py]: why not checking the order ?

File: addons/website_forum/tests/test_forum_tag.py [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        )
+        self.env['forum.tag'].flush_model()
+        self.assertEqual(self.forum.tag_most_used_ids, used_tag)
+        self.assertEqual(self.forum.tag_unused_ids, self.forum.tag_ids - used_tag)
+
+        self.assertEqual(self.base_forum.tag_most_used_ids, self.base_forum.tag_ids[:MOST_USED_TAGS_COUNT])
+        self.assertEqual(self.base_forum.tag_unused_ids, self.env['forum.tag'])
+
+    def test_tags_counts_most_used(self):
+        posts_per_tag = [(tag_id, post_count) for tag_id, post_count in zip(self.base_forum.tag_ids, range(2, 9))]
```
Comment (std-odoo) [inline addons/website_forum/tests/test_forum_tag.py]: the test will be even more robust if `range(2, 9)` is not sorted (so we double check the `posts_count DESC`)

File: addons/website_forum/models/forum_forum.py [hunk 19]
```diff
@@ -131,11 +135,46 @@ def _get_default_welcome_message(self):
...
+            order='forum_id, posts_count DESC, name, id',
+        )
+        current_forum_id = tags_data[0]['forum_id'][0]
+        current_most_used_count = 0
+        forum_tags = defaultdict(lambda: {'most_used_ids': [], 'unused_ids': []})
+
+        for tag_data in tags_data:
+            tag_id, tag_forum_id, posts_count = itemgetter('id', 'forum_id', 'posts_count')(tag_data)
+            if tag_forum_id[0] != current_forum_id:
+                current_most_used_count = 0
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: this is annoying to do that in python :/ but looks like it's not possible with the `read_group` (we can't choose the order when we do `:array_agg`)

I also tried in SQL, but the query is quite complicated, so not sure it's worth it

So to resume we either need to
- use `read_group` but sort in python -> mhe
- do a SQL query (couldn't do something simple, so big mhe)
- current implementation (order in sql thanks to `search_read`, and manually group in python) -> ok for now

(I just add it here to keep a trace if we change our mind)
```py
self.tag_ids.flush_recordset()
self.env.cr.execute("""
    SELECT f.id AS forum_id,
           ARRAY_AGG(DISTINCT mu.id) FILTER (WHERE mu.id IS NOT NULL) AS most_used,
           ARRAY_AGG(DISTINCT uu.id) FILTER (WHERE uu.id IS NOT NULL) AS unused
      FROM forum_forum AS f
JOIN LATERAL (SELECT t.id
              FROM forum_tag AS t
             WHERE t.forum_id = f.id AND t.posts_count != 0
          ORDER BY t.posts_count DESC, name, id
             LIMIT %(most_used)s
           ) AS mu ON TRUE
 LEFT JOIN forum_tag AS uu
        ON uu.forum_id = f.id AND uu.posts_count = 0
     WHERE f.id = ANY(%(forum_ids)s)
  GROUP BY f.id
""", {'forum_ids': self.ids, 'most_used': MOST_USED_TAGS_COUNT})
result = {
    line['forum_id']: (line['most_used'], line['unused'])
    for line in self.env.cr.dictfetchall()
}

for forum in self:
    most_used, unused = result.get(forum.id, ([], []))
    forum.tag_most_used_ids = self.env['forum.tag'].browse(most_used)
    forum.tag_unused_ids = self.env['forum.tag'].browse(unused)
```

File: addons/website_forum/models/forum_forum.py [hunk 20]
```diff
@@ -256,10 +295,13 @@ def _tag_to_write_vals(self, tags=''):
...
-    def get_tags_first_char(self):
-        """ get set of first letter of forum tags """
-        tags = self.env['forum.tag'].search([('forum_id', '=', self.id), ('posts_count', '>', 0)])
-        return sorted(set([tag.name[0].upper() for tag in tags if len(tag.name)]))
+    def _get_tags_first_char(self, tags=None):
+        """Get set of first letter of forum tags.
+
+        :param tags: tags recordset to further filter forum's tags that are also in these tags.
+        """
+        tag_ids = self.tag_ids if not tags else (self.tag_ids & tags)
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]:
```suggestion
        tag_ids = self.tag_ids if tags is not None else (self.tag_ids & tags)
```

**Review by std-odoo — COMMENTED** [review]
@flch-odoo Hi :) 

Next batch of comments :) 

I tried to optimize the computed method with the `read_group`, but we face some limitation of the ORM :/ I also posted my attempt to make a SQL query if it gives you ideas, but it seems quite complex so I think we can just keep the current implementation (except if we find a super simple query, but I doubt it's possible)

Cheers

File: addons/website_forum/models/forum_forum.py (hunk 20 above)
Comment (flch-odoo) [reply addons/website_forum/models/forum_forum.py]: `is None` :+1:

File: addons/website_forum/models/forum_tag.py (hunk 12 above)
Comment (flch-odoo) [reply addons/website_forum/models/forum_tag.py:62]: `base_domain`expects... a list of domains.
From the mixin:
```py
    def _search_get_detail(self, website, order, options):
        """...
        - base_domain: list of domains within which to perform the search
        ...
        """
```

File: addons/website_forum/tests/test_forum_tag.py (hunk 17 above)
Comment (flch-odoo) [reply addons/website_forum/tests/test_forum_tag.py]: I already did as it'sq the worst case scenario where high id  = high posts_count. 
But it's more explicit this way.

File: addons/website_forum/tests/test_forum_tag.py (hunk 18 above)
Comment (flch-odoo) [reply addons/website_forum/tests/test_forum_tag.py]: it is in the opposite of id`'s order. Do we really need more complexity?

File: addons/website_forum/tests/test_forum_tag.py (hunk 13 above)
Comment (flch-odoo) [reply addons/website_forum/tests/test_forum_tag.py:35]: I did something similar without extra import ;)

**Review by flch-odoo — COMMENTED** [review]
Thank you @std-odoo !
I've recovered from two additional bugs during this process, thanks for insisting on the order. One of the problem was in the order by of `forum_id` which only used sequence, so had mixed up results... (`forum_id.id` didn't help).
Also, I left the penultimate loop too early when we changed forum :grimacing: ...
I'm giving myself a few hours to come up with a query to use instead. I'll ping you on the task when I'm done either way.

File: addons/website_forum/tests/test_forum_tag.py (hunk 13 above)
Comment (std-odoo) [reply addons/website_forum/tests/test_forum_tag.py:35]: even better indeed :)

**Review by std-odoo — APPROVED** [review]

File: addons/website_forum/models/forum_forum.py [hunk 21]
```diff
@@ -131,11 +135,50 @@ def _get_default_welcome_message(self):
 
     can_moderate = fields.Boolean(string="Is a moderator", compute="_compute_can_moderate")
 
+    # tags
+    tag_ids = fields.One2many('forum.tag', 'forum_id', string='Tags')
+    tag_most_used_ids = fields.One2many('forum.tag', compute='_compute_tag_ids_usage')
+    tag_unused_ids = fields.One2many('forum.tag', compute='_compute_tag_ids_usage')
```
Comment (std-odoo) [inline addons/website_forum/models/forum_forum.py]: missing field label

**Review by std-odoo — COMMENTED** [review]
One last small thing spotted :) 

PS: the runbot is red about ci/style

**Comment by flch-odoo** [comment]
Merged in #123709

**Pull request closed without merging**